python/flare/code_grafter.py

```diff
@@ -0,0 +1,858 @@
+
+    Returns:
+        str: ASCII hexlified string of instruction opcode bytes for function.
+    """
+    header = ''
+    footer = ''
+    indent = ''
+    return _emit_fnbytes(_emit_ascii, header, footer, indent, fva, warn)
+
+
+def emit_fnbytes_python(fva=None, warn=True):
+    """Emit function bytes as Python code with disassembly in comments.
+
+    Args:
+        fva (numbers.Integral): function virtual address.
+            Defaults to here() if that is the start of a function, else
+            defaults to the start of the function that here() is a part of.
+        warn (bool): enable interactive warnings
+
+    Returns:
+        str: Python code you can spruce up and paste into a script.
+    """
+    header = 'instrs_{name} = (\n'
+    footer = ')'
+    indent = '    '
+    return _emit_fnbytes(_emit_for_python, header, footer, indent, fva, warn)
+
+
+def emit_fnbytes_c(fva=None, warn=True):
+    """Emit function bytes as C code with disassembly in comments.
+
+    Args:
+        fva (numbers.Integral): function virtual address.
+            Defaults to here() if that is the start of a function, else
+            defaults to the start of the function that here() is a part of.
+        warn (bool): enable interactive warnings
+
+    Returns:
+        str: C code you can spruce up and paste into a script.
+    """
+
+    header = 'unsigned char *instrs_{name} = {{\n'
+    footer = '};'
+    indent = '\t'
+    return _emit_fnbytes(_emit_for_c, header, footer, indent, fva, warn)
+
+
+def _emit_for_c(va, the_bytes, size):
+    disas = idc.GetDisasm(va)
+    byte_buf = ''.join(['\\x%s' % (binascii.hexlify(c)) for c in the_bytes])
+    return '"%s" /* %s */\n' % (byte_buf, disas)
+
+
+def _emit_for_python(va, the_bytes, size):
+    disas = idc.GetDisasm(va)
+    return "'%s' # %s\n" % (binascii.hexlify(the_bytes), disas)
+
+
+def _emit_ascii(va, the_bytes, size):
+    return binascii.hexlify(the_bytes)
+
+
+def _emit_fnbytes(emit_instr_cb, header, footer, indent, fva=None, warn=True):
+    """Emit function bytes in a format defined by the callback and
+    headers/footers provided.
+
+    Warns if any instruction operands are not consistent with
+    position-independent code, in which case the user may need to templatize
+    the position-dependent portions.
+    """
+    fva = fva or idc.here()
+    fva = idc.get_func_attr(fva, idc.FUNCATTR_START)
+    va_end = idc.get_func_attr(fva, idc.FUNCATTR_END)
+
+    # Operand types observed in position-independent code:
+    #   1: General Register (al,ax,es,ds...)
+    #   3: Base + Index
+    #   4: Base + Index + Displacement
+    #   5: Immediate
+    #   7: Immediate Near Address
+    # Notably missing because I want to note and handle these if/as they are
+    # encountered:
+    #   8: FPP register
+    #   9: 386 control register
+    #   10: 386 debug register
+    #   11: 386 trace register
+    optypes_position_independent = set([1, 3, 4, 5, 7])
```

Comment:
Prefer using the o_<name> for the operand type constants (like o_reg, o_mem) available as ida_ua.o_reg.